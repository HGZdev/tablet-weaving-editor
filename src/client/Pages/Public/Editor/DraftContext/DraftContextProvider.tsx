import React, {createContext, useState, ReactNode} from "react";
import {cloneDeep, isEqual} from "lodash";
import initDraft from "../__fixtures__/draft0.js";

export type DraftContextType = {
  draft: Draft;
  paletteOfColors: string[];
  selectedColor: number;
  setSelectedColor: (index: number) => void;
  onUploaded: (uplDraft: Draft) => void;
  onSkewToggle: (col: number) => void;
  onDirChange: (col: number, row: number, isTurnPoint: boolean) => void;
  onColorChange: (col: number, row: number) => void;
  onHolesChange: (newHoles: Draft["holes"]) => void;
  onPicksChange: (newPicks: Draft["picks"]) => void;
  onTabletsChange: (newTabletsNum: number) => void;
  onPaletteColorChange: (color: string) => void;
};

export const DraftContext = createContext<DraftContextType | undefined>(
  undefined
);

export const DraftProvider: React.FC<{children: ReactNode}> = ({children}) => {
  const [draft, setDraft] = useState<Draft>(initDraft as Draft);
  const [selectedColor, setSelectedColor] = useState<number>(0);
  const [paletteOfColors, setPaletteOfColors] = useState<string[]>(
    draft.paletteOfColors
  );

  const handleUploaded = (uplDraft: Draft) => {
    setDraft(uplDraft);
    setPaletteOfColors(uplDraft.paletteOfColors);
  };

  const handleSkewToggle = (col: number) => {
    const newSkews = draft.skews.slice() as typeof draft.skews;
    newSkews[col] = draft.skews[col] > 0 ? -1 : 1;
    if (!isEqual(draft.skews, newSkews))
      setDraft((d) => ({...d, skews: newSkews}));
  };

  const handleDirChange = (col: number, row: number, isTurnPoint: boolean) => {
    const newDirsChanges = cloneDeep(draft.dirsChanges);
    newDirsChanges[col][row] = isTurnPoint ? 0 : 1;
    if (!isEqual(draft.dirsChanges, newDirsChanges))
      setDraft((d) => ({...d, dirsChanges: newDirsChanges}));
  };

  const handleColorChange = (col: number, row: number) => {
    const newTablets = draft.tablets.slice() as typeof draft.tablets;
    newTablets[col][row] = paletteOfColors[selectedColor];
    setDraft((d) => ({...d, tablets: newTablets}));
  };

  const handleHolesChange = (newHoles: number) => {
    if (newHoles !== draft.holes) {
      const newTablets =
        newHoles > draft.holes
          ? (draft.tablets.map((t) => [
              ...t,
              t[t.length - 1],
            ]) as unknown as typeof draft.tablets)
          : (draft.tablets.map((t) =>
              t.slice(0, newHoles)
            ) as typeof draft.tablets);

      setDraft((d) => ({...d, holes: newHoles, tablets: newTablets}));
    }
  };

  const handlePicksChange = (newPicks: number) => {
    if (newPicks !== draft.picks) {
      const newDirsChanges =
        newPicks > draft.picks
          ? (draft.dirsChanges.map((t) => [
              ...t,
              0 as MakeDirChange,
            ]) as unknown as typeof draft.dirsChanges)
          : (draft.dirsChanges.map((t) =>
              t.slice(0, -1)
            ) as typeof draft.dirsChanges);

      setDraft((d) => ({...d, picks: newPicks, dirsChanges: newDirsChanges}));
    }
  };

  const handleTabletsChange = (newTabletsNum: number) => {
    if (!isEqual(newTabletsNum, draft.tablets?.length)) {
      const oneMore = newTabletsNum > draft.tablets?.length;
      const newDirsChanges = oneMore
        ? [...draft.dirsChanges, [...draft.dirsChanges.slice(-1)[0]]]
        : draft.dirsChanges.slice(0, -1);
      const newSkews = oneMore
        ? [...draft.skews, ...[...draft.skews.slice(-1)]]
        : draft.skews.slice(0, -1);
      const newTablets = oneMore
        ? [...draft.tablets, [...draft.tablets.slice(-1)[0]]]
        : draft.tablets.slice(0, -1);

      setDraft((d) => ({
        ...d,
        tablets: newTablets,
        skews: newSkews,
        dirsChanges: newDirsChanges,
      }));
    }
  };

  const handlePaletteColorChange = (color: string) => {
    const newPaletteOfColors = paletteOfColors.slice();
    newPaletteOfColors[selectedColor] = color;
    if (!isEqual(paletteOfColors, newPaletteOfColors))
      setPaletteOfColors(newPaletteOfColors);
  };

  return (
    <DraftContext.Provider
      value={{
        draft,
        paletteOfColors,
        selectedColor,
        setSelectedColor,
        onUploaded: handleUploaded,
        onSkewToggle: handleSkewToggle,
        onDirChange: handleDirChange,
        onColorChange: handleColorChange,
        onHolesChange: handleHolesChange,
        onPicksChange: handlePicksChange,
        onTabletsChange: handleTabletsChange,
        onPaletteColorChange: handlePaletteColorChange,
      }}
    >
      {children}
    </DraftContext.Provider>
  );
};
