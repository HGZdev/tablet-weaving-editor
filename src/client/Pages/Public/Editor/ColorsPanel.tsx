import {useState, useEffect} from "react";
import {ButtonPrimary} from "../../../Components/Buttons";
import {useDraft} from "./DraftContext/useDraft";

const ColorsPanel: React.FC = () => {
  const {
    paletteOfColors,
    selectedColor,
    setSelectedColor,
    onPaletteColorChange,
  } = useDraft();
  const [inputColor, setInputColor] = useState<ThreadColor>(
    paletteOfColors[selectedColor]
  );

  useEffect(() => {
    if (paletteOfColors[selectedColor] !== inputColor)
      setInputColor(paletteOfColors[selectedColor]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedColor]);

  return (
    <div title="colors-panel" className="flex flex-col gap-2">
      <div className="flex gap-1">
        {paletteOfColors?.map((color, i) => (
          <button
            key={i}
            title={color}
            className={`w-[1.5rem] h-[1.5rem] cursor-pointer ${
              selectedColor === i
                ? `border-2 border-gray-400 border-dashed`
                : `border-[0.5px] border-gray-400 border-solid`
            }`}
            style={{backgroundColor: color}}
            onClick={() => setSelectedColor(i)}
          />
        ))}
      </div>

      <div className="flex gap-2 w-56">
        <input
          title="color-input"
          type="text"
          className="input input-bordered rounded-none input-sm w-full"
          value={inputColor || ""}
          onChange={(e) => setInputColor(e.target.value)}
        />
        <ButtonPrimary
          type="submit"
          onClick={() => onPaletteColorChange(inputColor)}
        >
          Change color
        </ButtonPrimary>
      </div>
      {/* </FormikForm> */}
    </div>
  );
};

export default ColorsPanel;
