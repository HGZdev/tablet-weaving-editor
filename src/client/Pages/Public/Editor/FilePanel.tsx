import {useEffect, useRef, useState} from "react";
import styled from "styled-components";
import {downLoadFile, isJsonString} from "./helpers";
import {isEqual, pick} from "lodash";
import draft1 from "./__fixtures__/draft1";
import draft2 from "./__fixtures__/draft2";
import draft3 from "./__fixtures__/draft3";
import {ButtonPrimary, ButtonSecondary} from "../../../Components/Buttons";
import {useDraft} from "./DraftContext/useDraft";

const fieldsToSave: (keyof Draft)[] = [
  "fileName",
  "tablets",
  "holes",
  "picks",
  "skews",
  "dirsChanges",
  "paletteOfColors",
];

const HiddenUploadInput = styled.input`
  display: none;
`;

const FilePanel: React.FC = () => {
  const {draft, paletteOfColors, onUploaded} = useDraft();

  const inputFile = useRef<HTMLInputElement>(null);
  const [fileName, setFileName] = useState<string | undefined>(draft?.fileName);

  useEffect(() => {
    if (draft.fileName !== fileName) setFileName(draft.fileName);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [draft]);

  const handleUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    const fileReader = new FileReader();
    const rawFile = e.target.files?.[0];
    fileReader.readAsText(rawFile!, "UTF-8");
    fileReader.addEventListener("load", (e) => {
      const res = e.target?.result as string;
      const upload = isJsonString(res) && JSON.parse(res);
      if (upload && !isEqual(upload, draft)) {
        upload.fileName = rawFile?.name.replace(".json", "");
        onUploaded(upload);
      }
    });
  };

  const handleDownload = () => {
    const draftToSave = pick(
      {...draft, fileName, paletteOfColors},
      fieldsToSave
    );
    const draftStr = JSON.stringify(draftToSave);
    downLoadFile([draftStr], "application/json", `${fileName || "draft"}.json`);
  };

  const handleUploadClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();
    e.preventDefault();
    inputFile.current?.click();
  };

  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setFileName(e.target.value);

  const handleExampleDraftUpload = (draft: Draft) => onUploaded(draft);

  return (
    <div title="file-panel">
      <div className="flex gap-2">
        <div>
          <HiddenUploadInput
            className="hidden-file-upload-input"
            id={`file-input-${Date.now()}`}
            type="file"
            accept=".json"
            ref={inputFile}
            onChange={handleUpload}
          />
          <input
            className="file-upload-input input input-bordered input-sm rounded-none"
            type="text"
            name="File name"
            value={fileName}
            onChange={handleNameChange}
          />
        </div>
        <div className="flex gap-2">
          <ButtonSecondary
            className="file-upload-button"
            onClick={handleDownload}
          >
            Save project
          </ButtonSecondary>
          <ButtonPrimary onClick={handleUploadClick}>Open</ButtonPrimary>
        </div>
      </div>
      <div className="flex  gap-2 py-4">
        <span> Examples:</span>

        <button onClick={() => handleExampleDraftUpload(draft1 as Draft)}>
          draft 1
        </button>
        <button onClick={() => handleExampleDraftUpload(draft2 as Draft)}>
          draft 2
        </button>
        <button onClick={() => handleExampleDraftUpload(draft3 as Draft)}>
          draft 3
        </button>
      </div>
    </div>
  );
};

export default FilePanel;
