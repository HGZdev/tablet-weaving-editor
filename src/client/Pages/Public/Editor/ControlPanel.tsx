import ColorsPanel from "./ColorsPanel";
import FilePanel from "./FilePanel";
import InputsPanel from "./InputsPanel";

const ControlPanel: React.FC = () => {
  return (
    <div title="control-panel" className="flex flex-col h-full gap-4 p-4  ">
      <InputsPanel />
      <ColorsPanel />
      <FilePanel />
    </div>
  );
};

export default ControlPanel;
