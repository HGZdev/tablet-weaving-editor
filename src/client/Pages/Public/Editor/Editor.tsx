import React, {useRef} from "react";
import ControlPanel from "./ControlPanel";
import ThreadsPanel from "./ThreadsPanel";
import TabletsPanel from "./TabletsPanel";
import {FiX} from "react-icons/fi";
import {NavButton} from "../../../Components/Buttons";
import PageFrame from "../../../Components/PageFrame";
import useOverflowDetection from "../../../hooks.ts/useOverflowDetection";

const Sidebar: React.FC<{
  onToggleSidebar: () => void;
}> = ({onToggleSidebar}) => {
  return (
    <div className="flex flex-col items-start bg-base-100 h-full">
      {/* close sidebar button */}
      <NavButton onClick={onToggleSidebar} className="md:hidden">
        <FiX size={24} />
      </NavButton>
      <ControlPanel />
    </div>
  );
};

export const Main: React.FC = () => {
  const mainRef = useRef<HTMLDivElement>(null);
  const isOverflowing = useOverflowDetection(mainRef);

  return (
    <div
      ref={mainRef}
      className={`flex w-full ${
        isOverflowing ? "justify-start" : "justify-center"
      } overflow-x-auto pb-[15rem]`}
    >
      <div className="bg-white my-8 py-4 px-8 rounded-md">
        <ThreadsPanel />
        <TabletsPanel />
      </div>
    </div>
  );
};

const Editor: React.FC = () => {
  return <PageFrame {...{Sidebar, Main}} />;
};

export default Editor;
