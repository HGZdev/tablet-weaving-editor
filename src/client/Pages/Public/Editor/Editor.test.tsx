import {render} from "@testing-library/react";
import {beforeEach, describe, expect, test, vi} from "vitest";
import Editor from "./Editor";
import {
  findByText,
  findByTitle,
} from "../../../../tests/testing-library/helpers";
import {DraftProvider} from "./DraftContext/DraftContextProvider";
import {MemoryRouter} from "react-router-dom";

describe("Editor Component", () => {
  beforeEach(() => {
    const mockedTimestamp = new Date("2024-01-01T00:00:00Z").getTime();
    vi.spyOn(Date, "now").mockReturnValue(mockedTimestamp);
  });
  test("renders the Editor correctly", async () => {
    render(
      <MemoryRouter>
        <DraftProvider>
          <Editor />
        </DraftProvider>
      </MemoryRouter>
    );

    // Check if the heading is rendered
    expect(await findByText("Tablet Weaving Editor"));
    expect(await findByText("Change color"));

    expect(await findByTitle("tablets-panel")).toMatchSnapshot();
    expect(await findByTitle("threads-panel")).toMatchSnapshot();
    expect(await findByTitle("file-panel")).toMatchSnapshot();
    expect(await findByTitle("control-panel")).toMatchSnapshot();

    expect(await findByText("Save Project"));
  });
});
