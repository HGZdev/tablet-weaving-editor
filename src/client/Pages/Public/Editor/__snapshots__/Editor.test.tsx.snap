// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Editor Component > renders the Editor correctly 1`] = `
.c0 {
  padding: 0.1rem 0.2rem;
  width: 1.2rem;
  height: 1.7rem;
  cursor: pointer;
}

.c1 {
  padding: 0.1rem 0.2rem;
  width: 1.2rem;
  height: 1.7rem;
  border: 0.5px solid #fff;
  background-color: lightgrey;
  cursor: pointer;
}

.c2 {
  width: 100%;
  height: 100%;
  background-color: black;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     -20deg   );
}

.c3 {
  width: 100%;
  height: 100%;
  background-color: white;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     -20deg   );
}

.c4 {
  width: 100%;
  height: 100%;
  background-color: black;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     20deg   );
}

.c5 {
  width: 100%;
  height: 100%;
  background-color: white;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     20deg   );
}

<div
  class="inline-flex justify-center"
  title="tablets-panel"
>
  <div>
    <div
      class="c0"
    />
    <div
      class="c0 row-A"
      title="hole name"
    >
      <span>
        A
      </span>
    </div>
    <div
      class="c0 row-B"
      title="hole name"
    >
      <span>
        B
      </span>
    </div>
    <div
      class="c0 row-C"
      title="hole name"
    >
      <span>
        C
      </span>
    </div>
    <div
      class="c0 row-D"
      title="hole name"
    >
      <span>
        D
      </span>
    </div>
    <div
      class="c0"
    />
  </div>
  <div>
    <div
      class="c0 col-0"
      title="column number"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="c1 col-0 row-3 change-color"
      title="change color"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-2 change-color"
      title="change color"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-1 change-color"
      title="change color"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c1 col-0 row-0 change-color"
      title="change color"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c0 col-0 toggle-skew skew-S"
      title="toggle skew"
    >
      <span>
        S
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0 col-1"
      title="column number"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="c1 col-1 row-3 change-color"
      title="change color"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-2 change-color"
      title="change color"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c1 col-1 row-1 change-color"
      title="change color"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-0 change-color"
      title="change color"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c0 col-1 toggle-skew skew-S"
      title="toggle skew"
    >
      <span>
        S
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0 col-2"
      title="column number"
    >
      <span>
        3
      </span>
    </div>
    <div
      class="c1 col-2 row-3 change-color"
      title="change color"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-2 change-color"
      title="change color"
    >
      <div
        class="c5"
      />
    </div>
    <div
      class="c1 col-2 row-1 change-color"
      title="change color"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-0 change-color"
      title="change color"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c0 col-2 toggle-skew skew-Z"
      title="toggle skew"
    >
      <span>
        Z
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0 col-3"
      title="column number"
    >
      <span>
        4
      </span>
    </div>
    <div
      class="c1 col-3 row-3 change-color"
      title="change color"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-2 change-color"
      title="change color"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-1 change-color"
      title="change color"
    >
      <div
        class="c5"
      />
    </div>
    <div
      class="c1 col-3 row-0 change-color"
      title="change color"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c0 col-3 toggle-skew skew-Z"
      title="toggle skew"
    >
      <span>
        Z
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0"
    />
    <div
      class="c0 row-A"
      title="hole name"
    >
      <span>
        A
      </span>
    </div>
    <div
      class="c0 row-B"
      title="hole name"
    >
      <span>
        B
      </span>
    </div>
    <div
      class="c0 row-C"
      title="hole name"
    >
      <span>
        C
      </span>
    </div>
    <div
      class="c0 row-D"
      title="hole name"
    >
      <span>
        D
      </span>
    </div>
    <div
      class="c0"
    />
  </div>
</div>
`;

exports[`Editor Component > renders the Editor correctly 2`] = `
.c0 {
  padding: 0.1rem 0.2rem;
  width: 1.2rem;
  height: 1.7rem;
  cursor: pointer;
}

.c1 {
  padding: 0.1rem 0.2rem;
  width: 1.2rem;
  height: 1.7rem;
  border: 0.5px solid #fff;
  background-color: lightgrey;
  cursor: pointer;
}

.c2 {
  width: 100%;
  height: 100%;
  background-color: black;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     -20deg   );
}

.c3 {
  width: 100%;
  height: 100%;
  background-color: white;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     -20deg   );
}

.c4 {
  width: 100%;
  height: 100%;
  background-color: black;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     20deg   );
}

.c5 {
  width: 100%;
  height: 100%;
  background-color: white;
  border: 1px solid black;
  border-radius: 5px;
  transform: skew(     20deg   );
}

<div
  class="inline-flex g-4 justify-center"
  title="threads-panel"
>
  <div>
    <div
      class="c0"
    />
    <div
      class="c0 row-7"
      title="row number"
    >
      <span>
        8
      </span>
    </div>
    <div
      class="c0 row-6"
      title="row number"
    >
      <span>
        7
      </span>
    </div>
    <div
      class="c0 row-5"
      title="row number"
    >
      <span>
        6
      </span>
    </div>
    <div
      class="c0 row-4"
      title="row number"
    >
      <span>
        5
      </span>
    </div>
    <div
      class="c0 row-3"
      title="row number"
    >
      <span>
        4
      </span>
    </div>
    <div
      class="c0 row-2"
      title="row number"
    >
      <span>
        3
      </span>
    </div>
    <div
      class="c0 row-1"
      title="row number"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="c0 row-0"
      title="row number"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="c0"
    />
  </div>
  <div>
    <div
      class="c0 col-0"
      title="column number"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="c1 col-0 row-7 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-6 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-5 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c1 col-0 row-4 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-3 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-2 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-0 row-1 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c1 col-0 row-0 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c0 col-0 toggle-skew skew-S"
      title="toggle skew"
    >
      <span>
        S
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0 col-1"
      title="column number"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="c1 col-1 row-7 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-6 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c1 col-1 row-5 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-4 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-3 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-2 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c1 col-1 row-1 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c1 col-1 row-0 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c2"
      />
    </div>
    <div
      class="c0 col-1 toggle-skew skew-S"
      title="toggle skew"
    >
      <span>
        S
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0 col-2"
      title="column number"
    >
      <span>
        3
      </span>
    </div>
    <div
      class="c1 col-2 row-7 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-6 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c5"
      />
    </div>
    <div
      class="c1 col-2 row-5 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-4 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-3 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-2 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c5"
      />
    </div>
    <div
      class="c1 col-2 row-1 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-2 row-0 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c0 col-2 toggle-skew skew-Z"
      title="toggle skew"
    >
      <span>
        Z
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0 col-3"
      title="column number"
    >
      <span>
        4
      </span>
    </div>
    <div
      class="c1 col-3 row-7 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-6 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-5 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c5"
      />
    </div>
    <div
      class="c1 col-3 row-4 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-3 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-2 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c1 col-3 row-1 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c5"
      />
    </div>
    <div
      class="c1 col-3 row-0 toggle-rotation-direction dir-forward"
      title="toggle rotation direction"
    >
      <div
        class="c4"
      />
    </div>
    <div
      class="c0 col-3 toggle-skew skew-Z"
      title="toggle skew"
    >
      <span>
        Z
      </span>
    </div>
  </div>
  <div>
    <div
      class="c0"
    />
    <div
      class="c0 row-7"
      title="row number"
    >
      <span>
        8
      </span>
    </div>
    <div
      class="c0 row-6"
      title="row number"
    >
      <span>
        7
      </span>
    </div>
    <div
      class="c0 row-5"
      title="row number"
    >
      <span>
        6
      </span>
    </div>
    <div
      class="c0 row-4"
      title="row number"
    >
      <span>
        5
      </span>
    </div>
    <div
      class="c0 row-3"
      title="row number"
    >
      <span>
        4
      </span>
    </div>
    <div
      class="c0 row-2"
      title="row number"
    >
      <span>
        3
      </span>
    </div>
    <div
      class="c0 row-1"
      title="row number"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="c0 row-0"
      title="row number"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="c0"
    />
  </div>
</div>
`;

exports[`Editor Component > renders the Editor correctly 3`] = `
.c0 {
  display: none;
}

<div
  title="file-panel"
>
  <div
    class="flex gap-2"
  >
    <div>
      <input
        accept=".json"
        class="c0 hidden-file-upload-input"
        id="file-input-1704067200000"
        type="file"
      />
      <input
        class="file-upload-input input input-bordered input-sm rounded-none"
        name="File name"
        type="text"
        value="draft example"
      />
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="btn btn-sm  btn-secondary shadow-md shadow-gray-300 rounded-none file-upload-button "
        type="button"
      >
        Save project
      </button>
      <button
        class="btn btn-sm  btn-primary shadow-md shadow-gray-300 rounded-none  undefined"
        type="button"
      >
        Open
      </button>
    </div>
  </div>
  <div
    class="flex  gap-2 py-4"
  >
    <span>
       Examples:
    </span>
    <button>
      draft 1
    </button>
    <button>
      draft 2
    </button>
    <button>
      draft 3
    </button>
  </div>
</div>
`;

exports[`Editor Component > renders the Editor correctly 4`] = `
.c0 {
  display: none;
}

<div
  class="inline-flex flex-col gap-4 "
  title="control-panel"
>
  <h2>
    Control Panel
  </h2>
  <div
    class="flex flex-col gap-1"
    title="inputs-panel"
  >
    <div
      class="flex gap-2"
    >
      <label
        class="min-w-16"
      >
        Holes:
      </label>
      <input
        class="w-8 text-center"
        readonly=""
        title="holes-input"
        type="number"
        value="4"
      />
      <button
        class="btn btn-xs btn-neutral w-8 text-center"
        title="holes-input decrement"
      >
        <span>
          -
        </span>
      </button>
      <button
        class="btn btn-xs btn-neutral w-8 text-center"
        title="holes-input increment"
      >
        <span>
          +
        </span>
      </button>
    </div>
    <div
      class="flex gap-2"
    >
      <label
        class="min-w-16"
      >
        Picks:
      </label>
      <input
        class="w-8 text-center"
        readonly=""
        title="picks-input"
        type="number"
        value="8"
      />
      <button
        class="btn btn-xs btn-neutral w-8 text-center"
        title="picks-input decrement"
      >
        <span>
          -
        </span>
      </button>
      <button
        class="btn btn-xs btn-neutral w-8 text-center"
        title="picks-input increment"
      >
        <span>
          +
        </span>
      </button>
    </div>
    <div
      class="flex gap-2"
    >
      <label
        class="min-w-16"
      >
        Tablets:
      </label>
      <input
        class="w-8 text-center"
        readonly=""
        title="tablets-input"
        type="number"
        value="4"
      />
      <button
        class="btn btn-xs btn-neutral w-8 text-center"
        title="tablets-input decrement"
      >
        <span>
          -
        </span>
      </button>
      <button
        class="btn btn-xs btn-neutral w-8 text-center"
        title="tablets-input increment"
      >
        <span>
          +
        </span>
      </button>
    </div>
  </div>
  <div
    class="flex flex-col gap-2"
    title="colors-panel"
  >
    <div
      class="flex gap-1"
    >
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-2 border-gray-400 border-dashed"
        style="background-color: rgb(0, 0, 0);"
        title="#000000"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(192, 192, 192);"
        title="#C0C0C0"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(0, 0, 255);"
        title="#0000FF"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(0, 255, 0);"
        title="#00FF00"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(255, 255, 0);"
        title="#FFFF00"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(255, 165, 0);"
        title="#ffa500"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(255, 0, 0);"
        title="#FF0000"
      />
      <button
        class="w-[1.5rem] h-[1.5rem] cursor-pointer border-[0.5px] border-gray-400 border-solid"
        style="background-color: rgb(255, 255, 255);"
        title="#FFFFFF"
      />
    </div>
    <div
      class="flex gap-2 w-56"
    >
      <input
        class="input input-bordered rounded-none input-sm w-full"
        title="color-input"
        type="text"
        value="#000000"
      />
      <button
        class="btn btn-sm  btn-primary shadow-md shadow-gray-300 rounded-none  undefined"
        type="submit"
      >
        Change color
      </button>
    </div>
  </div>
  <div
    title="file-panel"
  >
    <div
      class="flex gap-2"
    >
      <div>
        <input
          accept=".json"
          class="c0 hidden-file-upload-input"
          id="file-input-1704067200000"
          type="file"
        />
        <input
          class="file-upload-input input input-bordered input-sm rounded-none"
          name="File name"
          type="text"
          value="draft example"
        />
      </div>
      <div
        class="flex gap-2"
      >
        <button
          class="btn btn-sm  btn-secondary shadow-md shadow-gray-300 rounded-none file-upload-button "
          type="button"
        >
          Save project
        </button>
        <button
          class="btn btn-sm  btn-primary shadow-md shadow-gray-300 rounded-none  undefined"
          type="button"
        >
          Open
        </button>
      </div>
    </div>
    <div
      class="flex  gap-2 py-4"
    >
      <span>
         Examples:
      </span>
      <button>
        draft 1
      </button>
      <button>
        draft 2
      </button>
      <button>
        draft 3
      </button>
    </div>
  </div>
</div>
`;
